src/string_impls/custom_parsers/hex.rs

```diff
@@ -0,0 +1,188 @@
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd. (admin@parity.io)
+// This file is a part of the scale-value crate.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//         http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{
+    stringify::{ParseError, ParseErrorKind},
+    Value,
+};
+
+/// Attempt to parse a hex string into a [`Value<()>`] (or more specifically,
+/// an unnamed composite).
+///
+/// - Returns an error if we see a leading `0x` and then see invalid hex
+///   characters after this.
+/// - Returns `None` if no `0x` is seen first.
+/// - Returns `Some(value)` if parsing was successful. In this case, the string
+///   reference given is wound forwards to consume what was parsed.
+pub fn parse_hex(s: &mut &str) -> Option<Result<Value<()>, ParseError>> {
+    let bytes = s.as_bytes();
+
+    // Need at least 2 bytes for 0x
+    if bytes.len() < 2 {
+        return None;
+    }
+
+    // Look for leading "0x"; None if this obviously isn't hex.
+    if bytes[0] != b'0' || bytes[1] != b'x' {
+        return None;
+    }
+
+    let mut composite_values = vec![];
+
+    // find all valid hex chars after 0x:
+    let mut idx = 2;
+    let mut last_nibble = None;
+    loop {
+        // Break if we hit end of string.
+        let Some(b) = bytes.get(idx) else {
+            break;
+        };
+
+        // Break as soon as we hit some non-alphanumeric char.
+        if !b.is_ascii_alphanumeric() {
+            break;
+        }
+
+        // 4 bit number from hex char
+        let hex_nibble = (b'a'..=b'f')
```

Comment:
this is a little hard to read and I'm a little bothered seeing another impl of this in parity repos but fine.

I think matching on the range is nicer here such as:

```rust
fn read_nibble(n: u8) -> Result<Option<u8>, Error> {
    match n {
        b'A'..=b'F' => n - b'A' + 10,
        b'a'..=b'f' => n - b'a' + 10,
        b'0'..=b'9' if n.is_ascii_alphanumeric() => n - b'0',
        b'0'..=b'9' => None,
        b => {
            return Some(Err(ParseErrorKind::custom(ParseHexError::InvalidChar(b as char)).at(idx)))
        }
    }
}
```

but it to you the code looks correct, if we don't want to support whitespace characters in the hex string :P